src/Language/Futhark/Syntax.hs

```diff
@@ -1317,3 +1340,269 @@ instance Pretty BinOp where
+bareCleanField (RecordFieldExplicit name e loc) =
+  RecordFieldExplicit name (bareCleanExp e) loc
+bareCleanField (RecordFieldImplicit name _ loc) =
+  RecordFieldImplicit name NoInfo loc
+
+bareCleanPat :: PatBase Info VName -> PatBase NoInfo VName
+bareCleanPat (TuplePat ps loc) = TuplePat (map bareCleanPat ps) loc
+bareCleanPat (RecordPat fs loc) = RecordPat (map (fmap bareCleanPat) fs) loc
+bareCleanPat (PatParens p loc) = PatParens (bareCleanPat p) loc
+bareCleanPat (Id v _ loc) = Id v NoInfo loc
+bareCleanPat (Wildcard _ loc) = Wildcard NoInfo loc
+bareCleanPat (PatAscription pat t loc) = PatAscription (bareCleanPat pat) (bareCleanTypeExp t) loc
+bareCleanPat (PatLit v _ loc) = PatLit v NoInfo loc
+bareCleanPat (PatConstr c _ ps loc) = PatConstr c NoInfo (map bareCleanPat ps) loc
+bareCleanPat (PatAttr attr p loc) = PatAttr attr (bareCleanPat p) loc
+
+bareCleanDimIndex :: DimIndexBase Info VName -> DimIndexBase NoInfo VName
+bareCleanDimIndex (DimFix e) =
+  DimFix $ bareCleanExp e
+bareCleanDimIndex (DimSlice x y z) =
+  DimSlice (bareCleanExp <$> x) (bareCleanExp <$> y) (bareCleanExp <$> z)
+
+bareCleanLoopForm :: LoopFormBase Info VName -> LoopFormBase NoInfo VName
+bareCleanLoopForm (For (Ident i _ loc) e) = For (Ident i NoInfo loc) (bareCleanExp e)
+bareCleanLoopForm (ForIn pat e) = ForIn (bareCleanPat pat) (bareCleanExp e)
+bareCleanLoopForm (While e) = While (bareCleanExp e)
+
+bareCleanCase :: CaseBase Info VName -> CaseBase NoInfo VName
+bareCleanCase (CasePat pat e loc) = CasePat (bareCleanPat pat) (bareCleanExp e) loc
+
+bareCleanSizeExp :: SizeExp Info VName -> SizeExp NoInfo VName
+bareCleanSizeExp (SizeExp e loc) = SizeExp (bareCleanExp e) loc
+bareCleanSizeExp (SizeExpAny loc) = SizeExpAny loc
+
+bareCleanTypeExp :: TypeExp Info VName -> TypeExp NoInfo VName
+bareCleanTypeExp (TEVar qn loc) = TEVar qn loc
+bareCleanTypeExp (TETuple tys loc) = TETuple (map bareCleanTypeExp tys) loc
+bareCleanTypeExp (TERecord fs loc) = TERecord (map (second bareCleanTypeExp) fs) loc
+bareCleanTypeExp (TEArray size ty loc) = TEArray (bareCleanSizeExp size) (bareCleanTypeExp ty) loc
+bareCleanTypeExp (TEUnique ty loc) = TEUnique (bareCleanTypeExp ty) loc
+bareCleanTypeExp (TEApply ty ta loc) = TEApply (bareCleanTypeExp ty) (bareCleanTypeArgExp ta) loc
+  where
+    bareCleanTypeArgExp (TypeArgExpSize size) =
+      TypeArgExpSize $ bareCleanSizeExp size
+    bareCleanTypeArgExp (TypeArgExpType tya) =
+      TypeArgExpType $ bareCleanTypeExp tya
+bareCleanTypeExp (TEArrow arg tya tyr loc) = TEArrow arg (bareCleanTypeExp tya) (bareCleanTypeExp tyr) loc
+bareCleanTypeExp (TESum cs loc) = TESum (map (second $ map bareCleanTypeExp) cs) loc
+bareCleanTypeExp (TEDim names ty loc) = TEDim names (bareCleanTypeExp ty) loc
+
+-- | Remove all annotations, parentesis, assertion and attribute from an expression, but retain the
+-- name/scope information.
+bareCleanExp :: ExpBase Info VName -> ExpBase NoInfo VName
```

Comment:
Is this only used in the normalisation pass?  If so, it should be defined there.

Reply:
It is currently used for SizeExpr comparison

Reply:
I would very much like to get rid of those instances (or rather, return them to the default).  If we need a bespoke way of comparing size expressions somewhere, then we should define an appropriate function rather than repurposing `==`.  Defining an `Eq` or `Ord` instance that is not straight structural equality is almost always a bad idea.

I just tried removing those instances entirely and the type checker at least still seems to work.

Reply:
It does work without them, but I think we need to check the normalised code to be sure that the same size is calculated twice, one as `n+1`, and as second time as `(n+1)` or with some other parentheses

Reply:
That would only be a constant number of re-evaluations, so no great tragedy.  And in any case - that equality definition should if necessary be part of the normalisation module.